k2/csrc/arcsort.cc

```diff
@@ -0,0 +1,67 @@
+// k2/csrc/arcsort.cc
+
+// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/arcsort.h"
+
+#include <algorithm>
+#include <numeric>
+#include <utility>
+#include <vector>
+
+#include "glog/logging.h"
+#include "k2/csrc/fsa.h"
+
+namespace k2 {
+void ArcSorter::GetSizes(Array2Size<int32_t> *fsa_size) {
+  CHECK_NOTNULL(fsa_size);
+  fsa_size->size1 = fsa_in_.size1;
+  fsa_size->size2 = fsa_in_.size2;
+}
+
+void ArcSorter::GetOutput(Fsa *fsa_out, int32_t *arc_map /*= nullptr*/) {
+  CHECK_NOTNULL(fsa_out);
+  CHECK_EQ(fsa_out->size1, fsa_in_.size1);
+  CHECK_EQ(fsa_out->size2, fsa_in_.size2);
+
+  using ArcWithIndex = std::pair<Arc, int32_t>;
+  std::vector<int32_t> indexes(fsa_in_.size2);  // arc index mapping
+  std::iota(indexes.begin(), indexes.end(), 0);
+  const auto arc_begin_iter = fsa_in_.data;
+  const auto index_begin_iter = indexes.begin();
+  int32_t num_states = fsa_in_.NumStates();
+  int32_t num_arcs = 0;
+  for (int32_t state = 0; state != num_states; ++state) {
+    fsa_out->indexes[state] = num_arcs;
+    int32_t begin = fsa_in_.indexes[state];
+    int32_t end = fsa_in_.indexes[state + 1];
+    std::vector<ArcWithIndex> arc_range_to_be_sorted;
+    arc_range_to_be_sorted.reserve(end - begin);
+    std::transform(arc_begin_iter + begin, arc_begin_iter + end,
```

Comment:
I don't think it's actually necessary to store the Arcs in this vector, though.  What I would do is simply sort a vector of integers, but use a comparator which had as a member the begin() of the original (input) arcs, and treats the integer args as indexes into that.

Reply:
... of course you'd still need a temporary place to put the arcs while copying, but it would require just 2 copies per arc, instead of as many as the sort requires.

Reply:
BTW, I'm not convinced that I would have made arc-sorting a non-in-place operation.  I don't see a scenario where you'd want to sort the arcs but still keep the original.

Reply:
That's the original interface you wrote. So could we change it to an in-place interface now as it seems there's no such scenario (non-in-place)?

Reply:
by saying `original interface`, I mean the non-in-place version for `ArcSort`